File: port/effs/sys_argv.js
// Sys.argv: the command line as one string: a flag per argument (`1` valid UTF-8, `0` not), then each argument, all
// separated by NUL. bun hands an invalid byte xx through as the lone surrogate U+DCxx, which Bend's strings refuse;
// the bytes are rebuilt from it and decoded as WHATWG does (TextDecoder), the same U+FFFD text as the C side (OQ-012).
function sys_argv_bytes(a) {
  const out = [];
  for (let i = 0; i < a.length; i += 1) {
    const c = a.charCodeAt(i);
    const paired = i > 0 && a.charCodeAt(i - 1) >= 0xD800 && a.charCodeAt(i - 1) <= 0xDBFF;
    if (c >= 0xDC80 && c <= 0xDCFF && !paired) {
      out.push(c - 0xDC00);
      continue;
    }
    const cp = a.codePointAt(i);
    if (cp > 0xFFFF) {
      i += 1;
    }
    for (const b of Buffer.from(String.fromCodePoint(cp), "utf8")) {
      out.push(b);
    }
  }
  return Uint8Array.from(out);
}

function sys_argv() {
  const lossy = new TextDecoder("utf-8");
  const strict = new TextDecoder("utf-8", { fatal: true });
  let flags = "";
  const texts = [];
  for (const a of cli_args) {
    const b = sys_argv_bytes(a);
    let ok = "1";
    try {
      strict.decode(b);
    } catch (e) {
      ok = "0";
    }
    flags += ok;
    texts.push(lossy.decode(b));
  }
  return [flags].concat(texts).join("\u0000");
}

File: port/effs/sys_argv.c
// Sys.argv: the command line as one string: a flag per argument (`1` valid UTF-8, `0` not), then each argument, all
// separated by NUL (which no argument can hold). io_str decodes the bytes as WHATWG does, so an invalid sequence
// reads as U+FFFD, as the original prints it; the flags let the core refuse an invalid VALUE the way the original
// does (S1.2, OQ-012). IO.args cannot: it hands the core the U+FFFD text and no way to tell it from a real one.
static int sys_argv_valid(const char* p, size_t n) {
  size_t need = 0;
  unsigned lo = 0x80, hi = 0xBF;
  for (size_t i = 0; i < n; i += 1) {
    unsigned b = (unsigned char)p[i];
    if (need > 0) {
      if (b < lo || b > hi) {
        return 0;
      }
      lo = 0x80;
      hi = 0xBF;
      need -= 1;
    } else if (b < 0x80) {
      continue;
    } else if (b < 0xC2 || b > 0xF4) {
      return 0;
    } else {
      need = b < 0xE0 ? 1 : b < 0xF0 ? 2 : 3;
      lo   = b == 0xE0 ? 0xA0 : b == 0xF0 ? 0x90 : 0x80;
      hi   = b == 0xED ? 0x9F : b == 0xF4 ? 0x8F : 0xBF;
    }
  }
  return need == 0;
}

Term sys_argv_run(Env e, Term* f, IoWork* w) {
  size_t len = (size_t)io_argc + 1;
  for (int i = 0; i < io_argc; i += 1) {
    len += strlen(io_argv[i]) + 1;
  }
  char* buf = malloc(len);
  if (buf == NULL) {
    return io_str(e, "", 0);
  }
  size_t n = 0;
  for (int i = 0; i < io_argc; i += 1) {
    buf[n++] = sys_argv_valid(io_argv[i], strlen(io_argv[i])) ? '1' : '0';
  }
  for (int i = 0; i < io_argc; i += 1) {
    size_t l = strlen(io_argv[i]);
    buf[n++] = 0;
    memcpy(buf + n, io_argv[i], l);
    n += l;
  }
  Term t = io_str(e, buf, (u64)n);
  free(buf);
  return t;
}

static void __attribute__((constructor)) sys_argv_use(void) {
  io_eff(CID_SYS_ARGV, sys_argv_run, 0);
}
